src/cppext.cpp

```diff
@@ -363,132 +351,73 @@ class dynamic_two_dim_matrix
     void next_row(int copy_to_column)
     {
         assert(copy_to_column < num_cols);
-        current_row++;
+        assert(!matrix.empty());
 
-        // Check whether we have to resize.
-        if (current_row == num_rows) {
-            resize();
-        }
+        std::vector<int32_t> new_row;
 
         // Copy values from the previous row.
         for (int j = 0; j <= copy_to_column; j++) {
-            matrix[current_row][j] = matrix[current_row - 1][j];
+            new_row.push_back(matrix.at(matrix.size() - 1).at(j));
         }
+        matrix.push_back(new_row);
+        rows_to_be_ignored.push_back(false);
     }
 
     /**
      * Indicate that the current row needs to be ignored.
      */
-    void ignore_current_row() { rows_to_be_ignored[current_row] = true; }
+    void ignore_current_row() { rows_to_be_ignored.back() = true; }
 
     /**
      * Delete the ignored rows from the matrix and shift the remaining rows upwards, i.e.,
      * if row \a is to be ignored, the new row \i of the matrix is row \i + 1;
      */
     void delete_ignored_rows()
     {
-        int num_ignored_so_far = 0;
-        for (int i = 0; i <= current_row; i++) {
-            if (rows_to_be_ignored[i]) {
-                num_ignored_so_far++;
-                delete[] matrix[i];
+        int i = 0;
+        for (auto it = matrix.begin(); it != matrix.end(); ) {
+            if (rows_to_be_ignored.at(i)) {
+                matrix.erase(it);
             } else {
-                matrix[i - num_ignored_so_far] = matrix[i];
+                ++it;
             }
+            ++i;
         }
-
-        // Add empty rows at the end. No need to go beyond current_row, as those are still untouched rows.
-        for (int i = current_row - num_ignored_so_far + 1; i <= current_row; i++) {
-            matrix[i] = initialized_array();
-        }
-        current_row -= num_ignored_so_far;
     }
 
     /**
      * Get the current matrix as a single dimensional array of size \a num_cols * (\a current_row + 1).
+     * Entries that were not added will get the default value.
      * @return the matrix as a single row array.
      */
-    int32_t* get_array()
+    int32_t* get_array(int32_t default_value)
     {
-        auto* array = new int32_t[num_cols * (current_row + 1)];
+        auto* array = new int32_t[num_cols * (matrix.size())];
```

Comment:
I am not sure about the rest of the code, but perhaps it is worthwhile to use `std::vector` instead of raw `int32_t*`:
`std::vector` knows about the size, it will automatically take care of correct deallocation, raw access through `vector::data()`, also it is cheap to `std::move` (it copies just a pointer+size), thus no loss of performance, yet readability is much better.

Reply:
Yeah, this is tricky. As long as the result would stay within the C++ world, `std::vector` would be prefered, I agree. For J-Ecdar purpose, we also need to port the results back into Java, which also needs some delicacy with the data types. At least using pure C pointers works for now.